ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,577 @@
+     */
+    final boolean afterDelete;
+
+    OperationType(boolean nonBlockingRouter, boolean afterDelete) {
+      this.nonBlockingRouter = nonBlockingRouter;
+      this.afterDelete = afterDelete;
+    }
+  }
+
+  /**
+   * Describes an operation chain and provides useful metadata and operation results for testing.
+   */
+  static class OperationChain {
+    final int chainId;
+    final BlobProperties properties;
+    final byte[] userMetadata;
+    final byte[] data;
+    final Queue<OperationType> operations;
+    final List<TestFuture> testFutures = new ArrayList<>();
+    final CountDownLatch latch = new CountDownLatch(1);
+    String blobId;
+
+    OperationChain(int chainId, BlobProperties properties, byte[] userMetadata, byte[] data,
+        Queue<OperationType> operations) {
+      this.chainId = chainId;
+      this.properties = properties;
+      this.userMetadata = userMetadata;
+      this.data = data;
+      this.operations = operations;
+    }
+  }
+
+  /**
+   * This class encapsulates a future and allows the user to define a check method that runs tests on
+   * the retrieved value in the future.
+   *
+   * @param <T> The type of the encapsulated future
+   */
+  private static abstract class TestFuture<T> {
+    final Future<T> future;
+    final String operationType;
+    final OperationChain opChain;
+
+    TestFuture(Future<T> future, String operationType, OperationChain opChain) {
+      this.future = future;
+      this.operationType = operationType;
+      this.opChain = opChain;
+    }
+
+    /**
+     * Generate a name for the tested operation
+     * @return the operation name
+     */
+    String getOperationName() {
+      return operationType + "-" + opChain.chainId;
+    }
+
+    /**
+     * Return the value inside the future or throw an {@link AssertionError} if an exception
+     * occurred.
+     * @return the value inside the future
+     */
+    T get() {
+      try {
+        return this.future.get();
+      } catch (Exception e) {
+        e.printStackTrace();
+        Assert.fail("Exception occured in operation: " + getOperationName() + ", exception: " + e);
+        return null;
+      }
+    }
+
+    /**
+     * Implement any testing logic here.
+     */
+    abstract void check();
+  }
+
+  /**
+   * A callback for router operations that starts the next operation in the chain after completion.
+   * The user can define a custom action on the result of the operation by overriding the {@code action()} method.
+   * @param <T> The callback's result type
+   */
+  private class TestCallback<T> implements Callback<T> {
+    final OperationChain opChain;
+    final boolean expectError;
+
+    TestCallback(OperationChain opChain, boolean expectError) {
+      this.opChain = opChain;
+      this.expectError = expectError;
+    }
+
+    @Override
+    public void onCompletion(T result, Exception exception) {
+      if (exception != null && !expectError) {
```

Comment:
I was using the callback to either submit the next operation in the chain asynchronously or end the operation chain if an exception is encountered or if there are no operations left in the chain.  I wanted assertions to be done from the main test thread so that is why I don't do much with the exception in the callback.